### PR TITLE
Recover ws stream errors and pack them into proper ServerErrors so internal errors are not leaked

### LINKED ISSUE #13680: better error on pruning when passing explicit offsets to WS endpoints
#13633 yields an `internal error` message from websockets on pruning.

https://github.com/digital-asset/daml/blob/02677d29cb1c30d93db084ad688345e9805bb283/ledger-service/http-json/src/itlib/scala/http/AbstractWebsocketServiceIntegrationTest.scala#L977-L991

Instead, we should emit a useful [still string] error message that indicates a pruning error. We don't have to indicate _which_ queries in a multi-query failed, just that the WS request failed for this reason.

This is probably due to gRPC exceptions not propagating upward properly to `Endpoints`, where they ought to go through the `EndpointsCompanion.Error.fromThrowable` mechanism that all requests' errors ought to go through.

* [ ] check whether we can improve this handling in `Endpoints` to direct such errors to the aforementioned error-mapping mechanism
* [ ] fix json-api documentation that mentions `internal error`

_Originally posted by @S11001001 in https://github.com/digital-asset/daml/issues/13506#issuecomment-1103089781_

### PR DESCRIPTION
**Pull request by realvictorprm**
Fixes #13680

### CODE REVIEW

**Comment by realvictorprm** [comment]
Sometimes fixes can be quite short and nice :)

**Review by ray-roestenburg-da — APPROVED** [review]

File: ledger-service/http-json/src/itlib/scala/http/AbstractWebsocketServiceIntegrationTest.scala [hunk 1]
```diff
@@ -979,16 +974,15 @@ abstract class AbstractWebsocketServiceIntegrationTest
...
-      t.closeReason should ===("internal error")
+    } yield inside(streamError) { case Seq(TextMessage.Strict(text)) =>
+      val errorResponse = decodeErrorResponse(text)
+      errorResponse.status shouldBe StatusCodes.BadRequest
+      errorResponse.ledgerApiError
+        .map { it =>
+          it.code shouldBe 9
+          it.message should include("PARTICIPANT_PRUNED_DATA_ACCESSED")
+        }
+        .getOrElse(fail("ledger api error was expected to be included in the response"))
```
Comment (S11001001) [inline ledger-service/http-json/src/itlib/scala/http/AbstractWebsocketServiceIntegrationTest.scala:985]: I do not think this should be a message on the stream with a successful close; it should be a stream failure as it was before.

File: ledger-service/http-json/src/failurelib/scala/http/FailureTests.scala [hunk 2]
```diff
@@ -390,9 +389,8 @@ abstract class FailureTests
         uri,
         query,
       ) via parseResp runWith respBefore(cid)).transform(x => Success(x))
-      _ = inside(r) { case Failure(e: PeerClosedConnectionException) =>
-        e.closeCode shouldBe 1011
-        e.closeReason shouldBe "internal error"
+      _ = inside(r) { case Success(_) =>
+        succeed
```
Comment (S11001001) [inline ledger-service/http-json/src/failurelib/scala/http/FailureTests.scala:393]: Likewise, the `closeReason` is the best place to report the new error.

**Comment by S11001001** [comment]
@realvictorprm I believe our best approach is to see if we can add that feature upstream ourselves, then, and block #13680 until it is released.

**Comment by S11001001** [comment]
> With upstream you are speaking of contributing to Akka and fixing it there, right?

Yes.

**Comment by S11001001** [comment]
> Do we want to get involved in longer discussions with the maintainers on the right implementation or would you proceed differently?

PRs have a way of cutting certain Gordian knots.

A few takeaways from some past free software contributions:

1. PRs will always be taken more seriously than issues or "me too"s.
2. An issue, or discussion of a fix, is a battle plan. A PR is a battle plan that is in contact with the enemy.
3. Maintainers are human, too; they might know the most about their project, but that doesn't mean contributors can't also contribute to their knowledge.

> there was no clear decision yet which implementation resolving the issue would most likely be accepted

I don't think such discussions can be interpreted so readily. There are a number of reasons such discussions take place. One of them is simply that no one has had sufficient motivation to actually try to implement a fix or feature.

Moreover, much of the complication discussed there deals with the implications of closing "in an orderly fashion", i.e. not losing buffered data. But _that is happening when the stream throws for error anyway, **is happening right now in the akka-http codebase**, and is also the only case that affects us._ The unstated implication of the discarding of buffered data is that _this is happening when streams throw right now and is considered acceptable in this case._

We don't need a comprehensive fix for akka/akka-http#2458, we only want to manipulate the details of cases that are failing right now. It would be perfectly fine, for example, if we could only report failure codes. It would be perfectly fine if we couldn't change the close code at all; we only need to specify a reason other than `internal error`.

I don't know for sure whether the subset issue we have is worth the broader discussion. It might be. But it might not be. For example:

1. What happens if you handle a certain `e` class [here](https://github.com/akka/akka-http/blob/90b5ecbc6f41b13af02153f3f61accb0458cf88e/akka-http-core/src/main/scala/akka/http/impl/engine/ws/FrameOutHandler.scala#L64-L67) specially?
2. Upstreaming that a bit, what if you detected it [here](https://github.com/akka/akka-http/blob/90b5ecbc6f41b13af02153f3f61accb0458cf88e/akka-http-core/src/main/scala/akka/http/impl/engine/ws/WebSocket.scala#L281) and emitted a new `BypassEvent` [introduced alongside this one](https://github.com/akka/akka-http/blob/90b5ecbc6f41b13af02153f3f61accb0458cf88e/akka-http-core/src/main/scala/akka/http/impl/engine/ws/FrameHandler.scala#L220)? (That's a little bigger change but maybe a little safer.)

These are just suggestions that are convenient for us, so that certainly distracts me from what would be best for akka-http. But I would feel silly if we could have done something like these or equally simple all along, but didn't because of some stated complications that turned out to not be so bad, or besides the point.

**Comment by realvictorprm** [comment]
Fine by me then :) I was simply unsure whether we are on agreement that this is too much of a possible time sink or not. Considering your answer this means it could fairly not be a time sink and work out well so let's do this :)

**Pull request closed without merging**